File: src/linear_algebra_vector_spaces_easy.py
"""
linear_algebra - vector_spaces (easy)
Generated: 2026-02-22T05:53:19.395428
"""

from problem_utils import *

def generate():
    problem_type = randint(1, 5)
    
    if problem_type == 1:
        # Check if a vector is in a given span (1000-1200)
        # Simple 2D case with obvious answer
        v1_vals = [nonzero(-3, 3), nonzero(-3, 3)]
        scalar = randint(2, 4)
        test_vec = [scalar * v1_vals[0], scalar * v1_vals[1]]
        
        v1 = Matrix(v1_vals)
        test = Matrix(test_vec)
        
        return problem(
            question=f"Is the vector ${latex(test)}$ in the span of ${latex(v1)}$?",
            answer=True,
            difficulty=(1000, 1150),
            topic="linear_algebra/vector_spaces",
            solution=steps(
                f"We need to check if there exists a scalar $c$ such that $c{latex(v1)} = {latex(test)}$",
                f"From the first component: $c \\cdot {v1_vals[0]} = {test_vec[0]}$, so $c = {scalar}$",
                f"Check second component: ${scalar} \\cdot {v1_vals[1]} = {scalar * v1_vals[1]} = {test_vec[1]}$ ✓",
                f"Therefore, the vector is in the span."
            ),
        )
    
    elif problem_type == 2:
        # Find a linear combination (1100-1250)
        c1 = randint(1, 3)
        c2 = randint(1, 3)
        v1_vals = [nonzero(-2, 2), nonzero(-2, 2)]
        v2_vals = [nonzero(-2, 2), nonzero(-2, 2)]
        
        result = [c1 * v1_vals[0] + c2 * v2_vals[0], c1 * v1_vals[1] + c2 * v2_vals[1]]
        
        v1 = Matrix(v1_vals)
        v2 = Matrix(v2_vals)
        result_vec = Matrix(result)
        
        return problem(
            question=f"Express ${latex(result_vec)}$ as a linear combination of ${latex(v1)}$ and ${latex(v2)}$. What is the coefficient of ${latex(v1)}$?",
            answer=c1,
            difficulty=(1100, 1250),
            topic="linear_algebra/vector_spaces",
            solution=steps(
                f"We want to find $a$ and $b$ such that $a{latex(v1)} + b{latex(v2)} = {latex(result_vec)}$",
                f"From the first component: $a \\cdot {v1_vals[0]} + b \\cdot {v2_vals[0]} = {result[0]}$",
                f"From the second component: $a \\cdot {v1_vals[1]} + b \\cdot {v2_vals[1]} = {result[1]}$",
                f"By inspection or solving: $a = {c1}$, $b = {c2}$",
                f"The coefficient of ${latex(v1)}$ is ${c1}$"
            ),
        )
    
    elif problem_type == 3:
        # Determine if vectors are linearly independent - simple 2D case (1150-1300)
        # Two parallel vectors (dependent)
        scalar = randint(2, 4)
        v1_vals = [nonzero(-3, 3), nonzero(-3, 3)]
        v2_vals = [scalar * v1_vals[0], scalar * v1_vals[1]]
        
        v1 = Matrix(v1_vals)
        v2 = Matrix(v2_vals)
        
        return problem(
            question=f"Are the vectors ${latex(v1)}$ and ${latex(v2)}$ linearly independent?",
            answer=False,
            difficulty=(1150, 1300),
            topic="linear_algebra/vector_spaces",
            solution=steps(
                f"Vectors are linearly independent if the only solution to $c_1{latex(v1)} + c_2{latex(v2)} = \\vec{{0}}$ is $c_1 = c_2 = 0$",
                f"Notice that ${latex(v2)} = {scalar}{latex(v1)}$",
                f"So ${scalar}{latex(v1)} - {latex(v2)} = \\vec{{0}}$ with non-zero coefficients",
                f"Therefore, the vectors are linearly dependent (not independent)."
            ),
        )
    
    elif problem_type == 4:
        # Add two vectors (1000-1100)
        v1_vals = [randint(-5, 5), randint(-5, 5)]
        v2_vals = [randint(-5, 5), randint(-5, 5)]
        
        v1 = Matrix(v1_vals)
        v2 = Matrix(v2_vals)
        result = v1 + v2
        
        return problem(
            question=f"Add the vectors ${latex(v1)} + {latex(v2)}$",
            answer=result,
            difficulty=(1000, 1100),
            topic="linear_algebra/vector_spaces",
            solution=steps(
                f"Add corresponding components:",
                f"First component: ${v1_vals[0]} + {v2_vals[0]} = {v1_vals[0] + v2_vals[0]}$",
                f"Second component: ${v1_vals[1]} + {v2_vals[1]} = {v1_vals[1] + v2_vals[1]}$",
                f"Result: ${latex(result)}$"
            ),
        )
    
    else:
        # Scalar multiplication (1000-1100)
        scalar = nonzero(-4, 4)
        v_vals = [randint(-4, 4), randint(-4, 4)]
        
        v = Matrix(v_vals)
        result = scalar * v
        
        return problem(
            question=f"Multiply the vector ${latex(v)}$ by the scalar ${scalar}$",
            answer=result,
            difficulty=(1000, 1100),
            topic="linear_algebra/vector_spaces",
            solution=steps(
                f"Multiply each component by ${scalar}$:",
                f"First component: ${scalar} \\cdot {v_vals[0]} = {scalar * v_vals[0]}$",
                f"Second component: ${scalar} \\cdot {v_vals[1]} = {scalar * v_vals[1]}$",
                f"Result: ${latex(result)}$"
            ),
        )

emit(generate())

File: src/linear_algebra_vector_spaces_easy.jl
# linear_algebra - vector_spaces (easy)
# Generated: 2026-03-08T21:10:21.834522

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x y z begin
    set_topic!("linear_algebra/vector_spaces")
    
    problem_type = rand(1:5)
    
    if problem_type == 1
        # Linear independence check for 2D vectors
        a = nonzero(-8, 8)
        b = nonzero(-8, 8)
        c = nonzero(-8, 8)
        d = nonzero(-8, 8)
        
        # Make them linearly independent or dependent
        if rand(Bool)
            # Independent: det ≠ 0
            while a*d - b*c == 0
                d = nonzero(-8, 8)
            end
            ans = "linearly independent"
            explanation = "The determinant is $(a*d - b*c) ≠ 0"
        else
            # Dependent: make second vector a multiple of first
            k = nonzero(-3, 3)
            c = k * a
            d = k * b
            ans = "linearly dependent"
            explanation = "\\vec{v}_2 = $(k)\\vec{v}_1"
        end
        
        problem(
            question="Determine if the vectors \\vec{v}_1 = \\begin{pmatrix} $(a) \\\\ $(b) \\end{pmatrix} and \\vec{v}_2 = \\begin{pmatrix} $(c) \\\\ $(d) \\end{pmatrix} are linearly independent or linearly dependent.",
            answer=ans,
            difficulty=(700, 950),
            answer_type="text",
            solution=steps(
                "Check if one vector is a scalar multiple of the other",
                "Calculate determinant: $(a)($(d)) - $(b)($(c)) = $(a*d - b*c)",
                explanation,
                "Answer: $(ans)"
            ),
            time=90
        )
        
    elseif problem_type == 2
        # Span of vectors in R^2
        a = nonzero(-6, 6)
        b = nonzero(-6, 6)
        c = randint(-8, 8)
        d = randint(-8, 8)
        
        # Check if (c,d) is in span
        if a*d - b*c == 0 && !(a == 0 && b == 0)
            # In span
            ans = "yes"
            if a != 0
                k = c // a
            else
                k = d // b
            end
            explanation = "We can write \\begin{pmatrix} $(c) \\\\ $(d) \\end{pmatrix} = $(k)\\begin{pmatrix} $(a) \\\\ $(b) \\end{pmatrix}"
        else
            ans = "no"
            explanation = "No scalar k exists such that k\\begin{pmatrix} $(a) \\\\ $(b) \\end{pmatrix} = \\begin{pmatrix} $(c) \\\\ $(d) \\end{pmatrix}"
        end
        
        problem(
            question="Is the vector \\begin{pmatrix} $(c) \\\\ $(d) \\end{pmatrix} in the span of \\begin{pmatrix} $(a) \\\\ $(b) \\end{pmatrix}? Answer 'yes' or 'no'.",
            answer=ans,
            difficulty=(750, 1000),
            answer_type="text",
            solution=steps(
                "A vector \\vec{w} is in span(\\vec{v}) if \\vec{w} = k\\vec{v} for some scalar k",
                "Check if $(c) = k($(a)) and $(d) = k($(b)) have a consistent solution",
                explanation,
                "Answer: $(ans)"
            ),
            time=120
        )
        
    elseif problem_type == 3
        # Dimension of span
        n_vectors = rand(2:3)
        
        if n_vectors == 2
            a1, b1 = nonzero(-7, 7), nonzero(-7, 7)
            
            if rand(Bool)
                # Make second independent
                a2, b2 = nonzero(-7, 7), nonzero(-7, 7)
                while a1*b2 - b1*a2 == 0
                    b2 = nonzero(-7, 7)
                end
                dim = 2
                explanation = "The vectors are linearly independent (determinant = $(a1*b2 - b1*a2) ≠ 0)"
            else
                # Make second dependent
                k = nonzero(-3, 3)
                a2, b2 = k*a1, k*b1
                dim = 1
                explanation = "The second vector is $(k) times the first, so they are linearly dependent"
            end
            
            problem(
                question="What is the dimension of the span of the vectors \\begin{pmatrix} $(a1) \\\\ $(b1) \\end{pmatrix} and \\begin{pmatrix} $(a2) \\\\ $(b2) \\end{pmatrix}?",
                answer=dim,
                difficulty=(800, 1100),
                solution=steps(
                    "The dimension equals the number of linearly independent vectors",
                    explanation,
                    "Answer: $(dim)"
                ),
                time=120
            )
        else
            # 3 vectors in R^3
            a1, b1, c1 = nonzero(-5, 5), randint(-5, 5), randint(-5, 5)
            a2, b2, c2 = randint(-5, 5), nonzero(-5, 5), randint(-5, 5)
            a3, b3, c3 = randint(-5, 5), randint(-5, 5), nonzero(-5, 5)
            
            # Make third a combination of first two sometimes
            if rand(Bool)
                k1, k2 = randint(-2, 2), randint(-2, 2)
                a3, b3, c3 = k1*a1 + k2*a2, k1*b1 + k2*b2, k1*c1 + k2*c2
                dim = 2
                explanation = "The third vector is a linear combination of the first two"
            else
                dim = 3
                explanation = "The three vectors are linearly independent"
            end
            
            problem(
                question="What is the dimension of the span of \\begin{pmatrix} $(a1) \\\\ $(b1) \\\\ $(c1) \\end{pmatrix}, \\begin{pmatrix} $(a2) \\\\ $(b2) \\\\ $(c2) \\end{pmatrix}, and \\begin{pmatrix} $(a3) \\\\ $(b3) \\\\ $(c3) \\end{pmatrix}?",
                answer=dim,
                difficulty=(900, 1200),
                solution=steps(
                    "Find the number of linearly independent vectors",
                    explanation,
                    "Answer: $(dim)"
                ),
                time=180
            )
        end
        
    elseif problem_type == 4
        # Subspace verification
        property = choice(["closed under addition", "closed under scalar multiplication", "contains zero vector"])
        
        a = nonzero(-8, 8)
        b = nonzero(-8, 8)
        
        if property == "contains zero vector"
            ans = "yes"
            explanation = "The zero vector \\begin{pmatrix} 0 \\\\ 0 \\end{pmatrix} satisfies $(a)(0) + $(b)(0) = 0"
            question_text = "Does the set \\{\\begin{pmatrix} x \\\\ y \\end{pmatrix} : $(a)x + $(b)y = 0\\} contain the zero vector? Answer 'yes' or 'no'."
        elseif property == "closed under addition"
            ans = "yes"
            explanation = "If $(a)x_1 + $(b)y_1 = 0 and $(a)x_2 + $(b)y_2 = 0, then $(a)(x_1+x_2) + $(b)(y_1+y_2) = 0"
            question_text = "Is the set \\{\\begin{pmatrix} x \\\\ y \\end{pmatrix} : $(a)x + $(b)y = 0\\} closed under addition? Answer 'yes' or 'no'."
        else
            ans = "yes"
            explanation = "If $(a)x + $(b)y = 0, then $(a)(kx) + $(b)(ky) = k($(a)x + $(b)y) = 0 for any scalar k"
            question_text = "Is the set \\{\\begin{pmatrix} x \\\\ y \\end{pmatrix} : $(a)x + $(b)y = 0\\} closed under scalar multiplication? Answer 'yes' or 'no'."
        end
        
        problem(
            question=question_text,
            answer=ans,
            difficulty=(850, 1150),
            answer_type="text",
            solution=steps(
                "Check the subspace property: $(property)",
                explanation,
                "Answer: $(ans)"
            ),
            time=150
        )
        
    else
        # Find a vector in the span
        a = nonzero(-6, 6)
        b = nonzero(-6, 6)
        k = randint(-4, 4)
        
        ans_x = k * a
        ans_y = k * b
        
        problem(
            question="Give an example of a vector in the span of \\begin{pmatrix} $(a) \\\\ $(b) \\end{pmatrix}. Write your answer as an ordered pair (x, y).",
            answer=fmt_tuple([ans_x, ans_y]),
            difficulty=(700, 900),
            answer_type="tuple",
            solution=steps(
                "A vector is in the span if it's a scalar multiple of the given vector",
                "Choose k = $(k)",
                "Answer: $(k) \\cdot \\begin{pmatrix} $(a) \\\\ $(b) \\end{pmatrix} = \\begin{pmatrix} $(ans_x) \\\\ $(ans_y) \\end{pmatrix}"
            ),
            time=90
        )
    end
end